adafruit_azureiot/iot_mqtt.py

```diff
@@ -0,0 +1,453 @@
+                    self._logger.error("ERROR: unknown twin! - {}".format(msg))
+        elif topic.startswith("devices/{}/messages/devicebound".format(self._device_id)):
+            self._handle_cloud_to_device_message(str(msg), topic)
+        else:
+            self._logger.error("ERROR: (unknown message) - {}".format(msg))
+
+    def _send_common(self, topic: str, data) -> None:
+        # Convert data to a string
+        if isinstance(data, dict):
+            data = json.dumps(data)
+
+        if not isinstance(data, str):
+            raise IoTError("Data must be a string or a dictionary")
+
+        self._logger.debug("Sending message on topic: " + topic)
+        self._logger.debug("Sending message: " + str(data))
+
+        retry = 0
+
+        while True:
+            gc.collect()
+            try:
+                self._logger.debug("Trying to send...")
+                self._mqtts.publish(topic, data)
+                self._logger.debug("Data sent")
+                break
+            except RuntimeError as runtime_error:
+                self._logger.info("Could not send data, retrying after 0.5 seconds: " + str(runtime_error))
+                retry = retry + 1
+
+                if retry >= 10:
+                    self._logger.error("Failed to send data")
+                    raise
+
+                time.sleep(0.5)
+                continue
+
+        print("finished _send_common")
+        gc.collect()
+
+    def _get_device_settings(self) -> None:
+        self._logger.info("- iot_mqtt :: _get_device_settings :: ")
+        self.loop()
+        self._send_common("$iothub/twin/GET/?$rid=0", " ")
+
+    # pylint: disable=R0913
+    def __init__(
+        self,
+        callback: IoTMQTTCallback,
+        wifi_manager: ESPSPI_WiFiManager,
+        hostname: str,
+        device_id: str,
+        key: str,
+        token_expires: int = 21600,
+        logger: logging = None,
+    ):
+        """Create the Azure IoT MQTT client
+        :param wifi_manager: The WiFi manager
+        :param IoTMQTTCallback callback: A callback class
+        :param str hostname: The hostname of the MQTT broker to connect to, get this by registering the device
+        :param str device_id: The device ID of the device to register
+        :param str key: The primary or secondary key of the device to register
+        :param int token_expires: The number of seconds till the token expires, defaults to 6 hours
+        :param adafruit_logging logger: The logger
+        """
+        self._callback = callback
+        self._wifi_manager = wifi_manager
+        self._mqtt_connected = False
+        self._auth_response_received = False
+        self._mqtts = None
+        self._device_id = device_id
+        self._hostname = hostname
+        self._key = key
+        self._token_expires = token_expires
+        self._username = "{}/{}/api-version={}".format(self._hostname, device_id, self._iotc_api_version)
```

Comment:
Does the API version update often enough to make `_iotc_api_version` a variable here?

Reply:
Changed to use the constant directly